### PR TITLE
fix: API usage alerting in production

### PR DESCRIPTION
**Pull request by zachaysan**
Thanks for submitting a PR! Please check the boxes below:

- [x] I have run [`pre-commit`](https://docs.flagsmith.com/platform/contributing#pre-commit) to check linting
- [x] I have added information to `docs/` if required so people know about the feature!
- [x] I have filled in the "Changes" section below?
- [x] I have filled in the "How did you test this code" section below?
- [x] I have used a [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) title for this Pull Request

## Changes

In order to address https://github.com/Flagsmith/flagsmith/issues/3461 two main issues were handled. First, an incorrectly implemented InfluxDB query was fixed in order to get the full API usage counts in a single scalar. Second, a local evaluation mode Flagsmith client was introduced in order to allowlist organisations into alerting so that any given mistake in the API alerting code won't disrupt too many users. 

## How did you test this code?

The API usage code was tested against the staging InfluxDB setup with a date range window set to about a year to gather the most amount of data. Thousands of organisations were tested without the same failure once the fix was implemented. 

The local evaluation Flagsmith client code was tested locally with settings variables set to appropriate values. During testing an issue with local evaluation cropped up where flags were not present if the method call happened before the local evaluation cache had been set, so this was temporarily worked around with a small sleep on the client's instantiation.

### CODE REVIEW

File: api/integrations/flagsmith/client.py [hunk 1]
```diff
@@ -21,14 +19,24 @@
...
 
-    if not _flagsmith_client:
-        _flagsmith_client = Flagsmith(**_get_client_kwargs())
+    try:
+        _flagsmith_client = _flagsmith_clients[name]
+    except (KeyError, TypeError):
+        kwargs = _get_client_kwargs()
+        kwargs["enable_local_evaluation"] = local_eval
+        _flagsmith_client = Flagsmith(**kwargs)
+        # TODO: Remove this once client has been fixed.
```
Comment (matthewelwell) [inline api/integrations/flagsmith/client.py]: Let's link the relevant PR.

File: api/tests/unit/organisations/test_unit_organisations_tasks.py [hunk 2]
```diff
@@ -247,6 +247,8 @@ def test_handle_api_usage_notifications_below_100(
         "organisations.tasks.get_current_api_usage",
     )
     mock_api_usage.return_value = 91
+    mocker.patch("organisations.tasks.get_client")
```
Comment (matthewelwell) [inline api/tests/unit/organisations/test_unit_organisations_tasks.py]: Should we test what happens when the flag is enabled / disabled?

Comment (zachaysan) [reply api/tests/unit/organisations/test_unit_organisations_tasks.py]: Sure I've added a new test for when the flag is disabled and reworked some of the mock logic to return the actual expected feature flag value.

File: api/integrations/flagsmith/client.py (hunk 1 above)
Comment (zachaysan) [reply api/integrations/flagsmith/client.py]: Done, although I'm sorta hoping we get that PR merged and deployed before this code is ready to merge.

Comment (matthewelwell) [reply api/integrations/flagsmith/client.py]: I've just released v3.6.0 with this in it.

Comment (zachaysan) [reply api/integrations/flagsmith/client.py]: Ok great I've removed the temporary shim.

Comment (matthewelwell) [reply api/integrations/flagsmith/client.py]: Well, I'm expecting to also see an updated poetry.lock to update the flagsmith client lib in this PR, right?

File: api/edge_api/identities/edge_identity_service.py [hunk 3]
```diff
@@ -15,4 +15,7 @@ def get_edge_identity_overrides(
             environment_id=environment_id, feature_id=feature_id
         )
     )
-    return [IdentityOverrideV2.parse_obj(item) for item in override_items]
+    for item in override_items:
+        item["environment_id"] = str(item["environment_id"])
+
+    return [IdentityOverrideV2.model_validate(item) for item in override_items]
```
Comment (khvn26) [inline api/edge_api/identities/edge_identity_service.py]: I'd avoid double iteration here:

```suggestion
    return [
        IdentityOverrideV2.model_validate(
            {**item, "environment_id": str(item["environment_id"])}
        )
        for item in override_items
    ]
```

File: api/integrations/flagsmith/client.py (hunk 1 above)
Comment (zachaysan) [reply api/integrations/flagsmith/client.py]: Great catch! I've added the update which came with a bunch of dependency upgrades, which pulled in an update to the `flagsmith-flag-engine` library which required a series of updates to a bunch of tests and some changes to the code that was using the `flagsmith-flag-engine` library.

File: api/util/mappers/dynamodb.py [hunk 4]
```diff
@@ -77,10 +77,12 @@ def map_environment_api_key_to_environment_api_key_document(
 def map_engine_identity_to_identity_document(
     engine_identity: "IdentityModel",
 ) -> Document:
-    return {
+    response = {
         field_name: _map_value_to_document_value(value)
         for field_name, value in engine_identity
     }
+    response["composite_key"] = engine_identity.composite_key
+    return response
```
Comment (matthewelwell) [inline api/util/mappers/dynamodb.py:85]: I'm a little unsure why this had to change @khvn26 @zachaysan ?

Comment (zachaysan) [reply api/util/mappers/dynamodb.py:85]: I believe it was an update to pydantic making computed properties no longer part of the generated document. @khvn26 had even opened an issue on their repo about it and the conclusion was that it was not going to get fixed at least until the first major upgrade.

**Review by matthewelwell — APPROVED** [review]

**Pull request merged**